### ISSUE TITLE
Discrepancy in the gamma gamma invariant mass distribution between Geant3 and Geant4

### ISSUE THREAD

**Issue by igjaegle** [JeffersonLab/HDGeant4#146]
Produced with
Software: central release except for halld_recon, halld_sim, and hdgeant4 (taken from master branch) 
Event generator: gen_pi0 (gamma + proton -> pi0 + proton -> gamma gamma + proton)
Launcher: local (but up-to-date) MCwrapper
Background: None
Simulation for GlueX run 30496 (LH2 target)

BCAL and FCAL responses are different in Geant3 & Geant4

Produced with ReactionFilter: 1_14__7_m14
![InvariantMass_Pi0_2](https://user-images.githubusercontent.com/22524742/76684908-050f6580-65e6-11ea-9b75-0dbe3af5859d.gif)

Monitoring histogram produced by FCAL_invmass plugin
![InvariantMass_Pi0_1](https://user-images.githubusercontent.com/22524742/76684909-05a7fc00-65e6-11ea-9ad6-19d1d77e8c32.gif)

Monitoring histogram produced by bcal_inv_mass plugin
![InvariantMass_Pi0_0](https://user-images.githubusercontent.com/22524742/76684910-05a7fc00-65e6-11ea-86cd-96de0a858e7b.gif)

Full bcal_invmass  histos for Geant4
![bcal_30496_g4](https://user-images.githubusercontent.com/22524742/76684953-591a4a00-65e6-11ea-864b-e7c31ee74cf2.gif)

Full bcal_invmass histos for Geant3
![bcal_30496_g3](https://user-images.githubusercontent.com/22524742/76684955-591a4a00-65e6-11ea-815f-9f90af760d17.gif)

**Comment by sdobbs** [comment]
Was hdgeant4 run in multithreaded mode?

**Comment by igjaegle** [comment]
I guess so, I used MCwrapper

**Comment by igjaegle** [comment]
yes in fact i used NCORES=4

**Comment by sdobbs** [comment]
Then you can check your MCWrapper configuration file to see if you
requested multiple cores.

If you did this is possibly a duplicate of issue #143

**Comment by igjaegle** [comment]
Was that not solved?

**Comment by jonzarling** [comment]
To my awareness the bug has not been tracked down.


It would be useful to check and see if running with 1 core looks different,
more in line with geant3. If so then is probably the same issue and we've
shown that it is reproducible with different generators.


Jon

**Comment by igjaegle** [comment]
Doing that now.

**Comment by sdobbs** [comment]
If the issue is open then it has not been resolved.

**Comment by igjaegle** [comment]
it is better but still different note Geant3 results did not change


![bcal_30496_g4](https://user-images.githubusercontent.com/22524742/76687084-114fee80-65f7-11ea-845b-3df8ccc90cc1.gif)
![InvariantMass_Pi0_2](https://user-images.githubusercontent.com/22524742/76687086-11e88500-65f7-11ea-95d4-80913860083e.gif)
![InvariantMass_Pi0_1](https://user-images.githubusercontent.com/22524742/76687087-11e88500-65f7-11ea-87b2-d586660852aa.gif)
![InvariantMass_Pi0_0](https://user-images.githubusercontent.com/22524742/76687088-11e88500-65f7-11ea-9a19-00ac23762ef0.gif)

**Comment by igjaegle** [comment]
If genuine, it means the showers are different in both simulation and presumably  the shower depth is different. Is there a plugin to look at the gammagamma inv mass for different bcal layers?

**Comment by rjones30** [comment]
Igal,

You are looking at a very high-level comparison. To learn more, you need to
study a specific geometry for the two photons. Depending on where exactly
the photon hits the bcal and at what angle, you will find differences in
the g3/g4 comparison. And even bigger differences between data and either
MC. That is where you should start. Studying longitudinal segmentation is a
higher-order issue.

-Richard Jones

-Richard Jones

**Comment by rjones30** [comment]
ps. You are looking for how and where the energy is escaping. It is not
from the back of the bcal, that much I can tell you. How and where the
energy escapes from the sides, and front of the modules mostly, and from
the ends for very forward showers is what you are looking for.
-Richard

**Comment by igjaegle** [comment]
tks, going to a single photon generator and shooting them at a fixed direction and energy will be more telling. Also note that not only BCAL is different but also FCAL.

**Comment by zisis0** [comment]
- right; energy that escapes the sides goes into adjacent modules, which
the clusterizer usually picks up ok.  
- We also have a shower albedo effect, where photons at small polar
angles "reflect" some of their energy across the BCAL and into
non-contiguous modules -- we have not quantified this or studied it in
detail, but likely those cells do not make it into a cluster and are
contributors to the unused energy.
- in the forward direction, stuff escapes (leakage) but also hits light
guides and SiPMs.  There a few physics processes there, but likely small
contributions.  Again, we have not simulated this.
-  In addition, shower energy and longitudinal development are
complicated things, for a longitudinally segmented calorimeter like the
BCAL.  Layers 1 and 2 should give more reliable results than layer 3
(particularly at lower Edep).  Karthik is looking into their behaviour.

Cheers, Zisis...




Dr. Zisis Papandreou
Head, Department of Physics
Professor of Physics, Ph.D, P.Phys.
---------------------------------------------------------
Department of Physics   
University of Regina        
3737 Wascana Parkway
Regina, SK  S4S 0A2     CANADA


Phone: (306) 585-5379







Fax: (306) 585-5659Email:  zisis@uregina.ca
Website:
https://www.uregina.ca/science/physics/people/faculty-research/zisis-papandreou/


>>> Richard Jones <notifications@github.com> 03/14/20 12:01 PM >>>

ps. You are looking for how and where the energy is escaping. It is not
from the back of the bcal, that much I can tell you. How and where the
energy escapes from the sides, and front of the modules mostly, and from
the ends for very forward showers is what you are looking for.
-Richard
 
On Sat, Mar 14, 2020 at 1:56 PM Richard Jones <rjones30@gmail.com>
wrote:
 
> Igal,
>
> You are looking at a very high-level comparison. To learn more, you
need
> to study a specific geometry for the two photons. Depending on where
> exactly the photon hits the bcal and at what angle, you will find
> differences in the g3/g4 comparison. And even bigger differences
between
> data and either MC. That is where you should start. Studying
longitudinal
> segmentation is a higher-order issue.
>
> -Richard Jones
>
> -Richard Jones
>
> On Sat, Mar 14, 2020 at 1:40 PM igjaegle <notifications@github.com>
wrote:
>
>> If genuine, it means the showers are different in both simulation and
>> presumably the shower depth is different. Is there a plugin to look
at the
>> gammagamma inv mass for different bcal layers?
>>
>> —
>> You are receiving this because you were assigned.
>> Reply to this email directly, view it on GitHub
>>
<https://github.com/JeffersonLab/HDGeant4/issues/146#issuecomment-599107083>,

**Comment by igjaegle** [comment]
I shoot 1, 2, 3, 4, 5, 6, and 7 GeV photon at two different polar angles, 5 deg and 50 deg i.e. into BCAL and FCAL, respectively.

Example for 1 GeV

![Gamma_bcal_5](https://user-images.githubusercontent.com/22524742/76706419-b9c88600-66bd-11ea-8eb8-7f315157958e.gif)
![Gamma_fcal_4](https://user-images.githubusercontent.com/22524742/76706420-b9c88600-66bd-11ea-9f4f-8e459ae640b4.gif)

There is a systematic difference between Geant3 and Geant4 at all energies. Geant4 is always peaking at the energy shooted and Geant3 always higher.

**Comment by igjaegle** [comment]
A possible solution for this difference between Geant3 and Geant4 is to have two different "gains" one for Geant3 and one for Geant4 both in BCAL and in FCAL.

**Comment by markdalton** [comment]
There is a parameter in hdgeant4 which has the purpose of making the energy scale the same between G3 and G4.  My analysis showed that this value is not correct.  See issue 113 where I study it for all angles and energies.  https://github.com/JeffersonLab/HDGeant4/issues/113
Unfortunately I wasn't able to convince the right people that this problem actually exists, so the fix has never been implemented.

**Comment by igjaegle** [comment]
Richard also mentions that offline and he showed me where to change these two parameters. I will fix it and search for an alternative solution.

**Comment by igjaegle** [comment]
The current fudge factors are: 1.015 for BCAL & 0.976 for FCAL. These factors are multiplied to the energy deposited in Geant4 to "match" Geant3 distribution. Note, we do not know if Geant3 is correct. 

FCAL photon energy fitted mean / photon energy thrown, the energy dependence correction is particularly bad below 2 GeV

![PhotonEnergyMean_fcal](https://user-images.githubusercontent.com/22524742/76775101-3f5e3b80-677b-11ea-8328-6e49dbac6bd3.gif)

Same for BCAL, here it seems no energy dependence correction is applied, note the fudge factor will no solve this problem. 

![PhotonEnergyMean_bcal](https://user-images.githubusercontent.com/22524742/76775102-3ff6d200-677b-11ea-901b-c2afa54970e1.gif)

**Comment by rjones30** [comment]
Igal, yes a single fudge factor will not solve the problem. This is also in
response to Mark Dalton, who wondered why his suggested fix was not
implemented. The answer is because the factor is energy and
angle-dependent, and also maybe position dependent along the length of the
module in that case of the bcal. This needs a more detailed treatment,
based on a more detailed study that takes into account energy, position,
and angle dependence.

-Richard Jones

**Comment by markdalton** [comment]
My analysis showed that taking a logarithmic average over the energy and angle in the BCAL there is a 2.5% downward overall shift in the energy for hdgeat4 compared to hdgeant.  Therefore I suggest that the BCAL number be increased from 1.015 to 1.040.  After that is done I will repeat my analysis to make sure that we hit the target.

**Comment by rjones30** [comment]
Mark,

A new PR is waiting on the hdgeant4 github repo with this change.

-Richard Jones

**Comment by markdalton** [comment]
Thanks.  It looks good.  I wanted to merge it but I couldn't.

On Mar 16, 2020, at 4:40 PM, Richard Jones <notifications@github.com<mailto:notifications@github.com>> wrote:


Mark,

A new PR is waiting on the hdgeant4 github repo with this change.

-Richard Jones

**Comment by rjones30** [comment]
Mark,

Ok, I just did it on your behalf. PR has been executed.

-Richard

**Comment by igjaegle** [comment]
Thank you Mark  Richard, I will update the FCAL during my shift tonight

**Comment by igjaegle** [comment]
For FCAL, the fudge factor is: 1.00372 and this is what I am getting for both BCAL and FCAL with the new fudges factors

![image](https://user-images.githubusercontent.com/22524742/76855131-d3391180-6826-11ea-9601-5e65fb88bb14.png)

**Comment by igjaegle** [comment]
After the change in BCAL & FCAL, the agreements between G3 & G4 is prety good.

![InvariantMass_Pi0_2](https://user-images.githubusercontent.com/22524742/76997186-40cd6680-6929-11ea-8774-410008b88c97.gif)

![InvariantMass_Pi0_1](https://user-images.githubusercontent.com/22524742/76997187-40cd6680-6929-11ea-9645-b5bf18f8d0b6.gif)

BCAL inv mass still shifted but comparable to G3
![c1](https://user-images.githubusercontent.com/22524742/76998095-dddccf00-692a-11ea-8416-7fcc1c776bbc.gif)

**Comment by markito3** [comment]
Wow. That does look good.
